website/client-old/js/services/undoService.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+angular.module('habitrpg')
+.factory('Undo', ['$rootScope', '$http', 'Tasks', 'Common',
+  function undo($rootScope, $http, Tasks, Common) {
+    var undoApi = {};
+
+    undoApi.history = [];
+
+    undoApi.addAction = function (action, data) {
+      undoApi.history.push({action: action, data: data});
+    };
+
+    undoApi.undoAction = function () {
+      var pop = undoApi.history.pop();
+      if (!pop) return;
+
+      // @TODO: We need to add a pop interface for each action
+      if (pop.action === 'scoreTask') {
+        var direction = pop.data.direction;
+
+        if (direction === "up")  {

```

Comment:
If you did decide to do this by storing the most recent change in an object, and if you wanted to allow just one level of undo, I think it would still be best to make the data structure an array containing the object, so that in future we could more easily expand it if we decided to allow multi-level undos.

That would also allow us to use the same data structure to provide a visible history of the user's recent actions (at some time in the future; not now). That's a feature request that we see fairly often.

To prevent the array becoming too large, it could be emptied on cron and/or when it had more than X objects in it then the oldest one could be deleted as the newest was added. 

@TheHollidayInn Would it be possible to add a few words to the PR's description to describe the extent of this feature? E.g., one level of undo or multiple levels; only tasks or tasks plus skills plus ...?  That will help Socialites and wiki editors be prepared for answering questions and writing documentation when it goes live, and it might help us in the future when we're looking back on changes.